spec/datadog/appsec/contrib/rack/integration_test_spec.rb

```diff
@@ -0,0 +1,284 @@
+        expect(spans).to have(1).items
+      end
+
+      describe 'GET request' do
+        subject(:response) { get url, params, headers }
+
+        let(:url) { '/success' }
+        let(:params) { {} }
+        let(:headers) { {} }
+
+        context 'with a non-event-triggering request' do
+          it { is_expected.to be_ok }
+
+          it_behaves_like 'a GET 200 span'
+          it_behaves_like 'a trace with AppSec tags'
+          it_behaves_like 'a trace without AppSec events'
+        end
+
+        context 'with an event-triggering request in headers' do
+          let(:headers) { { 'HTTP_USER_AGENT' => 'Nessus SOAP' } }
+
+          it { is_expected.to be_ok }
+          it { expect(triggers).to be_a Array }
+
+          it_behaves_like 'a GET 200 span'
+          it_behaves_like 'a trace with AppSec tags'
+          it_behaves_like 'a trace with AppSec events'
+        end
+
+        context 'with an event-triggering request in query string' do
+          let(:params) { { q: '1 OR 1;' } }
+
+          it { is_expected.to be_ok }
+
+          it_behaves_like 'a GET 200 span'
+          it_behaves_like 'a trace with AppSec tags'
+          it_behaves_like 'a trace with AppSec events'
+        end
+
+        context 'with an event-triggering request in IP' do
+          skip 'TODO: config not implemented'
+
+          # TODO: let(:config) { { ip_denylist: ['1.2.3.4'] } }
+          let(:headers) { { 'HTTP_X_FORWARDED_FOR' => '1.2.3.4' } }
+
+          it { is_expected.to be_ok }
+
+          # TODO: it_behaves_like 'a GET 403 span'
+          it_behaves_like 'a trace with AppSec tags'
+          # TODO: it_behaves_like 'a trace with AppSec events'
+        end
+
+        context 'with an event-triggering response' do
+          let(:url) { '/admin.php' } # well-known scanned path
+
+          it { is_expected.to be_not_found }
+          it { expect(triggers).to be_a Array }
+
+          it_behaves_like 'a GET 404 span'
+          it_behaves_like 'a trace with AppSec tags'
+          it_behaves_like 'a trace with AppSec events'
+        end
+      end
+
+      describe 'POST request' do
+        subject(:response) { post url, params, headers }
+
+        let(:url) { '/success' }
+        let(:params) { {} }
+        let(:headers) { {} }
+
+        context 'with a non-event-triggering request' do
+          it { is_expected.to be_ok }
+
+          it_behaves_like 'a POST 200 span'
+          it_behaves_like 'a trace with AppSec tags'
+          it_behaves_like 'a trace without AppSec events'
+        end
+
+        context 'with an event-triggering request in application/x-www-form-url-encoded body' do
+          let(:params) { { q: '1 OR 1;' } }
+
+          let(:middlewares) do
+            [
+              Datadog::Tracing::Contrib::Rack::TraceMiddleware,
+              Datadog::AppSec::Contrib::Rack::RequestMiddleware,
+              Datadog::AppSec::Contrib::Rack::RequestBodyMiddleware,
+            ]
+          end
```

Comment:
I'm not sure I understand the exact scope of your question 😅 

`Datadog::AppSec::Contrib::Rack::RequestBodyMiddleware` is the one needed to handle parameters coming from a request body. Note that `request.env['rack.request.form_hash']` is lazily filled and empty until `request.POST` is called, so that middleware ends up helping to ensure that.

In a real app it may need to be placed in an arbitrary location, whereas `Datadog::Tracing::Contrib::Rack::TraceMiddleware` and `Datadog::AppSec::Contrib::Rack::RequestMiddleware` should be placed near or at the top.

It is inconsequential to other test cases, and it also documents how to use these in the field, so I opted not to have a kitchensink "one size fits all" app that would add this middleware always (this use case would be best served by apps in `integration/apps`)

Reply:
My question is, being this an integration spec, it seems a bit weird to manually be modifying the middlewares. Is this something that customers will need to do to their own apps? If so, I guess the spec is realistic. But if not, can we use the same mechanism here, perhaps?

Reply:
> Is this something that customers will need to do to their own apps?

Yes, in the Rack case the users build their own Rack stack, and we can't divine what it is so we can't do it automatically as with Rails or Sinatra, so they need to add our middlewares manually at the appropriate place.

In the Sinatra case, if they want JSON body parsing they're going to manually add `rack-contrib`'s `JSONBodyParser`.

So in both cases, setting up the Rack middleware stack appropriately for relevant tests is representative of what users will do.